Replace modelica:// by modelica:/

HansOlsson opened the pull request:
Fix the major problem that using authority as a path is problematic, since they are normally case-insensitive.
(And keep the old variant as a deprecated feature to not break compatibility.)

This should be straightforward to adapt in tools, and then in libraries - in a safe way without breaking compatibility.

henrikt-ma reviewed (changes requested):
We shouldn't miss the opportunity to make the new Modelica URIs more useful than the ones we have today.  That's the purpose of MCP #2663.

HansOlsson commented:
> We shouldn't miss the opportunity to make the new Modelica URIs more useful than the ones we have today. That's the purpose of MCP #2663.

We should correct Modelica URIs so that they work right now; and can be updated with minimal changes in models.
Restructuring all resources as proposed in #2663 would break compatibility in various ways and is thus problematic and shouldn't block this good change.

henrikt-ma commented:
> We should correct Modelica URIs so that they work right now; and can be updated with minimal changes in models.
> Restructuring all resources as proposed in #2663 would break compatibility in various ways and is thus problematic and shouldn't block this good change.

No, #2663 isn't breaking backwards compatibility thanks to the use of a new form of Modelica URIs that can't be mistaken for the old form. Hence, the problem is that if we do what is suggested here, we will miss the opportunity to introduce a better mechanism in a backwards compatible way.

HansOlsson commented:
I have now found the specific part that requires the authority to be case-insensitive, which allows things like modelica://modelica.mechanics/ with current syntax.

It's found in RFC3986 (but likely not in the original RFC1738) https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2 so technically it's not the authority that is case-insensitive, but its host-part (which amounts to the same). (Which means that modelica URIs were broken when it was added in 2009, since the rfc is from 2005.)

henrikt-ma reviewed (changes requested):
Nothing has changes since my last review, but it would be nice to try to find a solution after we're done releasing 3.6.

henrikt-ma commented:
Considering that we use the period for separating identifiers of the class, the leading path separator in `modelica:/Modelica.Mechanics/Resources/foo.txt` doesn't really make sense to me.  Since it doesn't really make sense, we should also consider omitting it, like this:
```
modelica:Modelica.Mechanics/Resources/foo.txt
```

HansOlsson commented:
> Considering that we use the period for separating identifiers of the class, the leading path separator in `modelica:/Modelica.Mechanics/Resources/foo.txt` doesn't really make sense to me. Since it doesn't really make sense, we should also consider omitting it, like this:
> 
> ```
> modelica:Modelica.Mechanics/Resources/foo.txt
> ```

The leading path separator has semantic meaning in URIs, and the leading path separator makes sense currently - and allow us to later introduce a variant without - with different semantics.

HansOlsson commented:
> The leading path separator has semantic meaning in URIs, and the leading path separator makes sense currently - and allow us to later introduce a variant without - with different semantics.

Thus I think we should move forward with this one as is.

henrikt-ma reviewed (changes requested):
I'm missing a description of the deprecated use of the host part to give the class name.

HansOlsson commented:
> I'm missing a description of the deprecated use of the host part to give the class name.

It was sort of there, but I cleaned it up, made it into a separate paragraph, and added two examples.

henrikt-ma commented on chapters/packages.tex:
```diff
@@ -278,34 +278,40 @@ \subsection{The within Clause}\label{the-within-clause}
 scheme or when restructuring packages that resource references resolve
 to the same resource). Any storage scheme should be constrained such
 that a resource with a given path should be unique for any package name
-that precedes it. The first part of the path shall not be the name of a
-class in the package given by the authority.
```
Why was this removed rather than reformulated for the new form of Modelica-URI?

HansOlsson replied:
Will re-add.

henrikt-ma reviewed (changes requested):
Looks OK with the exception of removing the constraint on the "first part of the path".

henrikt-ma commented:
It would also be good to see the conflicts resolved before giving a final approval.

HansOlsson commented:
> It would also be good to see the conflicts resolved before giving a final approval.

Now done. It was the comma-change that caused the conflict.

henrikt-ma reviewed (approved):
Looks good now.
